docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,216 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# New Storage Interface
+
+## Motivation
+
+The current storage interface was evolved from the legacy etcd2 implementation. It has a few issues that prevents us from further optimizing the storage path efficiency while keeping the code maintainable and readable.
+
+We have several goals for the new storage design. Once we reach these goals, the storage layer should be a completely decoupled component from other core logic parts of Kubernetes. That way we can make it easier to maintain without affecting any external code.
+
+This list of goals we are trying to achieve with this proposal:
+
+- maintainability: The storage code should be straightforward and easy to understand. The storage layer should have very clear responsibility boundary. It should not require huge amount of effort for people to understand other parts of Kubernetes in order to start coding and fixing issues in storage layer.
+
+- testability: The storage code should be >90% (now it is ~70%) test covered. Storage developers should not rely on external packages to test the functionality of the storage layer. Users should feel confident in the correctness of the implementation as long as the unit/integration tests pass.
+
+- clear API: The storage API should be minimalistic and with clear guarantees.  Users of storage pkg should see the guarantees immediately from the methods naming and related docs. For example, we should have guarantees between Read and Write. Should the user expect to read the value at least as new as the previous local write? Or should the user expect to read the value at least as new as the current global write?
+
+- CPU and memory efficient: The storage layer should be designed to support object indexing and watch indexing. Storage layer should be designed to hide all storage related optimization details. The storage should be the single place to optimize server storage. If an index is built, users should expect to select objects in reasonable amount of time (for example < O(logN) for list). To optimize caching size of pod objects, we should do all the work in storage system without touching the pod object defined at API layer.
```

Comment:
This is just an example of what we could do here. Pod passed in from the API layer does not have to have the exact same memory representation in storage cache.